src/modules/ejector/ejector.py

```diff
@@ -0,0 +1,211 @@
+from src.providers.consensus.typings import Validator
+from src.services.exit_order import ValidatorToExitIterator
+from src.typings import BlockStamp, EpochNumber
+from src.utils.abi import named_tuple_to_dataclass
+from src.utils.validator_state import is_active_validator
+from src.web3py.extentions.lido_validators import LidoValidator, NodeOperatorGlobalIndex
+from src.web3py.typings import Web3
+
+
+logger = logging.getLogger(__name__)
+
+
+class Ejector(BaseModule, ConsensusModule):
+    """
+    Module that ejects lido validators depends on withdrawal requests stETH value.
+
+    Flow:
+    1. Calculate withdrawals amount to cover with ETH.
+    2. Calculate ETH rewards prediction per epoch.
+    3. Calculate withdraw epoch for next validator
+    Loop:
+        a. Calculate predicted rewards we get until we reach withdraw epoch
+        b. Check if validators to eject + predicted rewards + current balance is enough to finalize withdrawal requests
+            - If True - eject all validators in list. End.
+        c. Get next validator to eject.
+        d. Recalculate withdraw epoch
+
+    4. Decode lido validators into bytes and send report transaction
+    """
+    CONSENSUS_VERSION = 1
+    CONTRACT_VERSION = 1
+
+    AVG_EXPECTING_WITHDRAWALS_SWEEP_DURATION_MULTIPLIER = 0.5
+
+    def __init__(self, w3: Web3):
+        self.report_contract = w3.lido_contracts.validators_exit_bus_oracle
+        super().__init__(w3)
+
+        self.prediction_service = RewardsPredictionService(w3)
+
+    def execute_module(self, blockstamp: BlockStamp):
+        report_blockstamp = self.get_blockstamp_for_report(blockstamp)
+        if report_blockstamp:
+            self.process_report(report_blockstamp)
+
+    @lru_cache(maxsize=1)
+    def build_report(self, blockstamp: BlockStamp) -> tuple:
+        validators = self.get_validators_to_eject(blockstamp)
+
+        data, data_format = encode_data(validators)
+
+        return ReportData(
+            self.CONSENSUS_VERSION,
+            blockstamp.ref_slot,
+            len(validators),
+            data_format,
+            data,
+        ).as_tuple()
+
+    def get_validators_to_eject(self, blockstamp: BlockStamp) -> list[tuple[NodeOperatorGlobalIndex, LidoValidator]]:
+        chain_config = self._get_chain_config(blockstamp)
+
+        rewards_speed_per_epoch = self.prediction_service.get_rewards_per_epoch(blockstamp, chain_config)
+        epochs_to_sweep = self._get_sweep_delay_in_epochs(blockstamp)
+
+        to_withdraw_amount = self.get_total_unfinalized_withdrawal_requests_amount(blockstamp)
+        total_current_balance = self._get_total_balance(blockstamp)
+
+        validators_to_eject = []
+        validator_to_eject_balance_sum = 0
+
+        validators_iterator = ValidatorToExitIterator(
+            w3=self.w3,
+            blockstamp=blockstamp,
+            c_conf=chain_config,
+        )
+
+        for validator in validators_iterator:
```

Comment:
need to limit validators with `maxValidatorExitRequestsPerReport` from `OracleReportSanityChecker - getOracleReportLimits`